cdc/log.cc

```diff
@@ -538,23 +563,136 @@ class transformer final {
+                                    // maps and lists are just flattened
+                                    [&] (const collection_type_impl& type) {
+                                        std::vector<std::pair<bytes_view, bytes_view>> result;
+                                        process([&](const bytes_view& key, const bytes_view& value, bool live) {
+                                            result.emplace_back(key, value);
+                                        });
+                                        return map_type_impl::serialize_partially_deserialized_form(result, cql_serialization_format::internal());
+                                    },
+                                    // set need to transform from mutation view
+                                    [&] (const set_type_impl& type) {
+                                        std::vector<bytes_view> result;
+                                        process([&](const bytes_view& key, const bytes_view& value, bool live) {
+                                            result.emplace_back(key);
+                                        });
+                                        return type.serialize_partially_deserialized_form(result, cql_serialization_format::internal());
+                                    },
+                                    // for user type we collect the fields in the mutation and set to
+                                    // tuple of value or tuple of null in case of delete.
+                                    // fields not in the mutation are null in the enclosing tuple, signifying "no info"
+                                    [&](const user_type_impl& type) {
+                                        std::vector<bytes_opt> res(type.size());
+                                        std::array<bytes_view_opt, 1> tmp;
+                                        // type of the actual value in log column tuple.
+                                        auto res_type = static_pointer_cast<const tuple_type_impl>(dst->type)->type(1);
+                                        auto tt = static_pointer_cast<const tuple_type_impl>(res_type);
+                                        process([&](const bytes_view& key, const bytes_view& value, bool live) {
+                                            auto idx = deserialize_field_index(key);
+                                            tmp[0] = live ? bytes_view_opt{value} : std::nullopt;
+                                            res[idx].emplace(static_pointer_cast<const tuple_type_impl>(tt->all_types()[idx])->build_value(tmp));
+                                        });
+                                        return tt->build_value(res);
+                                    },
+                                    [&] (const abstract_type& o) -> bytes {
+                                        throw std::runtime_error(format("collection_mutation_view::is_any_live: unknown type {}", o.name()));
```

Comment:
forgot to change the error message

Reply:
It is a great error message and I support it. But I will also change it.